src/rez/package_order.py

```diff
@@ -608,6 +609,265 @@ def from_pod(cls, data):
+                            v = str(v)
+                        v = VersionRange(v)
+                new_versions.append(v)
+            if num_false > 1:
+                raise ConfigurationError("version_priorities for CustomPackageOrder may "
+                                         "only have one False / empty value")
+            parsed_dict[package] = new_versions
+        return parsed_dict
+
+    def to_pod(self):
+        return dict(packages=self._packages_to_pod(self.packages_dict))
+
+    @classmethod
+    def from_pod(cls, data):
+        return cls(packages=data["packages"])
+
+
+class PyPAPackageOrder(PackageOrder):
+    """A package order that allows for package ordering according to PEP440/PyPA.
+
+    A prerelease argument can be specified to order prerelease versions in front
+    of release versions
+
+    For example, given the versions [1.0b2, 1.0a1, 1.0, 1.1, 1.0rc1, 1.1b2],
+    an orderer initialized with ``prerelease=""`` would give the order:
+     [1.1, 1.1b2, 1.0, 1.0rc1, 1.0b2, 1.0a1].
+    an orderer initialized with ``prerelease="b"`` would give the order:
+     [1.1b2, 1.1, 1.0b2, 1.0rc1, 1.0, 1.0a1].
+
+    """
+    name = "pypa"
+
+    def __init__(self, prerelease: Literal["", "rc", "b", "a"] = "", packages=None):
+        super().__init__(packages)
+        self.prerelease = prerelease
+
+    @staticmethod
+    def get_pypa_sort_key(version: Version, prerelease: Literal["", "rc", "b", "a"] = ""):
+        """
+        Get a sort key for sorting Versions by PyPA rules.
+
+        The prerelease argument allows for risk tolerance to be set, such that we can opt into
+        allowing preview/rc versions ahead of release versions.
+        """
+        import rez.vendor.packaging.version as pypa
+        pypa_version = pypa.parse(str(version))
+
+        key = pypa_version._key
+        if not isinstance(pypa_version, pypa.Version):
+            # Fallback to pypa legacy sorting if this isn't a compatible version format.
+            return key
+
+        # We want to allow prerelease versions up to a specific level above release versions.
+        # If the prerelease token is not set, then we will sort release versions to the front.
+        # To do this, we are going to replace the "pre" component of the sort key tuple
+        # provided by packaging.
+        key_list = list(key)
+        # The packaging key comprises (epoch, release, pre, post, dev, local)
+        pre = key_list[2]
+
+        if pre in (pypa.Infinity, -pypa.Infinity):
+            # Keep the relative order the same for packages that have no prerelease, but
+            # Make sure the first key forces them to be after our prerelease preference.
+            risk_allowance_token = pypa.Infinity
+        elif not prerelease:
+            # We have no risk tolerance, so sort prereleases to the bottom.
+            risk_allowance_token = -pypa.Infinity
+        elif pre >= (prerelease, ):
```

Comment:
is `pre` a tuple?   would be nice to have a type annotation for the variable above.